clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+                | '0'..='9'
+                | '_'
+                | '-'
+                | '!'
+                | '?'
+                | '+'
+                | '<'
+                | '>'
+                | '='
+                | '/'
+                | '*' => ident.push(self.next),
+                _ => {
+                    if is_separator(self.next) {
+                        return ident;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_trait_identifier(&mut self) -> String {
+        let start_line = self.last_line as u32;
+        let start_column = self.last_column as u32;
+        let mut ident = String::new();
+
+        loop {
+            match self.next {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' => ident.push(self.next),
+                '>' => return ident,
+                _ => {
+                    if is_separator(self.next) {
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line: self.line as u32,
+                                start_column: self.column as u32,
+                                end_line: self.line as u32,
+                                end_column: self.column as u32,
+                            },
+                            e: LexerError::ExpectedClosing('>'),
+                        });
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line,
+                                start_column,
+                                end_line: start_line,
+                                end_column: start_column,
+                            },
+                            e: LexerError::NoteToMatchThis('<'),
+                        });
+                        return ident;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharTraitIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_principal(&mut self) -> String {
+        let mut principal = String::new();
+
+        loop {
+            self.read_char();
+            match self.next {
+                'A'..='Z' | '0'..='9' => principal.push(self.next),
+                _ => {
+                    if is_separator(self.next) {
+                        return principal;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharPrincipal(self.next));
+                        return principal;
+                    }
+                }
+            }
+        }
+    }
+
+    fn read_unsigned(&mut self) -> u128 {
```

Comment:
I don't think this code does what it's supposed to do if there is ever a literal unsigned number that is greater than `u128::MAX`.  If this happens, then the whole contract should fail analysis.  It looks like this code instead attempts to convert the textual representation into a `u128` without handling the overflow case.

Reply:
I think such an occurence should cause the lexer to log an error at this location?

Reply:
More of a design-level question: I see that the lexer will return a `String` for every other token, and rely on the parser to try to convert the `String` into a Clarity `Value` of the corresponding type.  The parser, not the lexer, identifies malformed tokens.  Should that also be happening here, for `uint` and `int`?

Reply:
Yes, that makes sense. The lexer should just be verifying the allowed characters are used and the parser should validate the complete token. I will make that change.

Reply:
Fixed in a43454bd2

Reply:
Rebased, now 639e4030e348158f63c0b8fbacbb52083b95fb74